reconcile/openshift_namespaces.py

```diff
@@ -1,127 +1,134 @@
-                    init_projects=True)
-    ob.init_specs_to_fetch(
-        ri,
-        oc_map,
-        namespaces=namespaces,
-        override_managed_types=['Namespace']
-    )
+def get_desired_state(
+        namespaces: List[Dict[str, str]]) -> List[Dict[str, str]]:
 
     desired_state = []
-    for cluster, namespace, _, _ in ri:
-        if cluster not in oc_map.clusters():
-            continue
-        desired_state.append({"cluster": cluster, "namespace": namespace})
+    for namespace in namespaces:
+        state = NS_STATUS_PRESENT
+        if namespace.get("delete"):
+            state = NS_STATUS_ABSENT
+
+        desired_state.append({"cluster": namespace["cluster"]["name"],
+                              "namespace": namespace["name"],
+                              "state": state})
+
+    return desired_state
 
-    return oc_map, desired_state
 
+def get_shard_namespaces() -> List[Dict[str, str]]:
+    all_namespaces = queries.get_namespaces()
+    namespaces = {}
+    to_be_ignored = []
+    for namespace in all_namespaces:
+        shard_key = f'{namespace["cluster"]["name"]}/{namespace["name"]}'
+        if is_in_shard(shard_key):
+            if shard_key not in namespaces:
+                namespaces[shard_key] = namespace
+            else:
+                to_be_ignored.append(shard_key)
+    for shard_key in to_be_ignored:
+        logging.debug(
+            f"Found multiple definitions for the namespace {shard_key};"
+            " Ignoring")
+        del namespaces[shard_key]
+
+    return namespaces.values()
 
-def check_ns_exists(spec, oc_map):
+
+def check_ns_exists(spec: Dict[str, str], oc_map: OC_Map) -> bool:
     cluster = spec['cluster']
     namespace = spec['namespace']
-
     try:
-        create = not oc_map.get(cluster).project_exists(namespace)
-        return spec, create
+        exists = oc_map.get(cluster).project_exists(namespace)
+        return spec, exists
     except StatusCodeError as e:
-        msg = 'cluster: {},'
-        msg += 'namespace: {},'
-        msg += 'exception: {}'
-        msg = msg.format(cluster,
-                         namespace,
-                         str(e))
+        msg = (
+            f'cluster: {cluster}, namespace: {namespace}, exception: {str(e)}'
+        )
         logging.error(msg)
 
     return spec, None
 
 
-def create_new_project(spec, oc_map):
+def manage_projects(spec: Dict[str, str], oc_map: OC_Map, action: str) -> None:
     cluster = spec['cluster']
     namespace = spec['namespace']
 
     oc = oc_map.get(cluster)
     if not oc:
         logging.log(level=oc.log_level, msg=oc.message)
         return None
-    oc.new_project(namespace)
+
+    if action == NS_ACTION_CREATE:
+        try:
+            oc.new_project(namespace)
+        except StatusCodeError as e:
+            msg = (
+                f'cluster: {cluster}, namespace: {namespace}, '
+                f'exception: {str(e)}'
+            )
+            logging.error(msg)
```

Comment:
we'd need to indicate that there was an error to make the integration do it's best but eventually fail

Reply:
Not sure if I understand what you mean. I see this error message format is common in a lot of integrations. Any idea or example on what you have in mind?

Reply:
He returns the error code and lets `run` do `sys.exit` as in all other integrations. This looks solid to me.

Reply:
indeed, we just need the integration to exit with an error when this happens.